File: receipt-analyzer-service/src/main/java/pl/madzierski/daniel/app/receipt/scan_resolver/service/PDFService.java
package pl.madzierski.daniel.app.receipt.scan_resolver.service;

import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.rendering.PDFRenderer;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

@Service
public class PDFService {

    private final String tmpFilePath;

    public PDFService(@Value("${pdf-service.tmp-dir}") String tmpFilePath) {
        this.tmpFilePath = tmpFilePath;
    }

    public List<String> dividePdfFileToImages(String pdfFilePath) {
        if (pdfFilePath == null || pdfFilePath.trim().isEmpty()) {
            throw new IllegalArgumentException("Requirement failed");
        }
        try {
            File pdfFile = new File(pdfFilePath);
            List<String> outputPaths = new ArrayList<>();
            try (PDDocument document = PDDocument.load(pdfFile)) {
                PDFRenderer renderer = new PDFRenderer(document);
                int numberOfPages = document.getNumberOfPages();
                for (int page = 0; page < numberOfPages; page++) {
                    BufferedImage image = renderer.renderImageWithDPI(page, 300f);
                    File outputFilePath = new File(tmpFilePath + "/" + UUID.randomUUID() + "/" + page + ".png");
                    outputFilePath.getParentFile().mkdirs();
                    ImageIO.write(image, "png", outputFilePath);
                    outputPaths.add(outputFilePath.getAbsolutePath());
                }
            }
            return outputPaths;
        } catch (Exception e) {
            e.printStackTrace();
            return Collections.emptyList();
        }
    }
}

File: receipt-analyzer-service/src/main/kotlin/pl/madzierski/daniel/app/receipt/scan_resolver/service/PDFService.kt
package pl.madzierski.daniel.app.receipt.scan_resolver.service

import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.rendering.PDFRenderer
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.awt.image.BufferedImage
import java.io.File
import java.util.*
import javax.imageio.ImageIO

@Service
class PDFService(
    @Value("\${pdf-service.tmp-dir}") private val tmpFilePath: String
) {

    fun dividePdfFileToImages(pdfFilePath: String): List<String> {
        require(pdfFilePath.isNotBlank())
        return try {
            val pdfFile = File(pdfFilePath)
            val document = PDDocument.load(pdfFile)
            val renderer = PDFRenderer(document)
            (0 until document.numberOfPages).map { page ->
                val image: BufferedImage = renderer.renderImageWithDPI(page, 300F)
                val outputFilePath = File("$tmpFilePath/${UUID.randomUUID()}/${page}.png")
                outputFilePath.parentFile.mkdirs()
                ImageIO.write(image, "png", outputFilePath)
                outputFilePath.absolutePath
            }
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }
}